Speedup heapsort by 1.8x by making it branchless

Voultapher opened the pull request:
`slice::sort_unstable` will fall back to heapsort if it repeatedly fails to find a good pivot. By making the core child update code branchless it is much faster. On Zen3 sorting 10k `u64` and forcing the sort to pick heapsort, results in:

~455us -> 278us~

455us -> 249us

Voultapher committed:
Speedup heapsort by 1.5x by making it branchless

`slice::sort_unstable` will fall back to heapsort if it repeatedly fails to find
a good pivot. By making the core child update code branchless it is much faster.
On Zen3 sorting 10k `u64` and forcing the sort to pick heapsort, results in:

455us -> 278us

Voultapher commented:
r? thomcc

scottmcm commented on library/core/src/slice/sort.rs:
```diff
@@ -198,9 +198,7 @@ where
-            if child + 1 < v.len() && is_less(&v[child], &v[child + 1]) {
-                child += 1;
-            }
+            child += (child + 1 < v.len() && is_less(&v[child], &v[child + 1])) as usize;
```
Hmm, the `&&` here is still a branch, and a necessary one, so maybe there's an alternative way to write this to emphasize the parts?  Maybe something like
```suggestion
            if child + 1 < v.len() {
                // We need a branch to be sure not to out-of-bounds index,
                // but it's highly predictable.  The comparison, however,
                // is better done branchless, especially for primitives.
                child += is_less(&v[child], &v[child + 1]) as usize;
            }
```

Voultapher replied:
Great point, I really like your suggestion. All three generate different x86-64 assembly, I'll benchmark it tomorrow. Based on the assembly I suspect your suggestion will perform the best.

- Original https://godbolt.org/z/M9zrMWbW1
- Combined cond https://godbolt.org/z/7Es3avoYn
- Split cond https://godbolt.org/z/nhxoKcW3z

scottmcm replied:
Oh, nice, the `child += is_less(…, …)` version gives a `cmp`+`adc`.

Diff, in case it's helpful for anyone else, since I made it anyway: <https://godbolt.org/z/arcsxcEc5>.

Voultapher replied:
I just tested it and it drops it down to 249, neat.

Voultapher committed:
Split branches in heapsort child selection

This allows even better code-gen, cmp + adc. While also more clearly
communicating the intent.

Voultapher commented:
r? @scottmcm

scottmcm commented:
Thanks for benching it!

@bors r+

(Ok to rollup because this is in the *fallback* sorting/selecting path, so is unlikely to matter for compiler perf.)

bors commented:
:pushpin: Commit ee0376c368c50c7dadc84801e88cfdbf250b92a4 has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm ee0376c368c50c7dadc84801e88cfdbf250b92a4 -->
<!-- homu: {"type":"Approved","sha":"ee0376c368c50c7dadc84801e88cfdbf250b92a4","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

thomcc commented:
@bors rollup=never

thomcc commented:
@scottmcm Not sure I agree — that's true but compiler is huge and it's hard to say what matters. Better safe than sorry IMO.

scottmcm commented:
@thomcc Sure, no objections, especially since the queue isn't too long right now.

bors commented:
:hourglass: Testing commit ee0376c368c50c7dadc84801e88cfdbf250b92a4 with merge b7089e0dd3e988270f34f182d3749ea5fff5a18f...
<!-- homu: {"type":"BuildStarted","head_sha":"ee0376c368c50c7dadc84801e88cfdbf250b92a4","merge_sha":"b7089e0dd3e988270f34f182d3749ea5fff5a18f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4154495167/jobs/7188054760)
Approved by: scottmcm
Pushing b7089e0dd3e988270f34f182d3749ea5fff5a18f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"scottmcm","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4154495167/jobs/7188054760"},"merge_sha":"b7089e0dd3e988270f34f182d3749ea5fff5a18f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4154495167/jobs/7188054760)
Approved by: scottmcm
Pushing b7089e0dd3e988270f34f182d3749ea5fff5a18f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"scottmcm","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4154495167/jobs/7188054760"},"merge_sha":"b7089e0dd3e988270f34f182d3749ea5fff5a18f"} -->

rust-timer commented:
Finished benchmarking commit (b7089e0dd3e988270f34f182d3749ea5fff5a18f): [comparison URL](https://perf.rust-lang.org/compare.html?start=d0940161283216ade541946474c759621b6997c5&end=b7089e0dd3e988270f34f182d3749ea5fff5a18f&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d0940161283216ade541946474c759621b6997c5&end=b7089e0dd3e988270f34f182d3749ea5fff5a18f&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d0940161283216ade541946474c759621b6997c5&end=b7089e0dd3e988270f34f182d3749ea5fff5a18f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.6%  | [1.6%, 1.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.3% | [-3.3%, -3.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d0940161283216ade541946474c759621b6997c5&end=b7089e0dd3e988270f34f182d3749ea5fff5a18f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
